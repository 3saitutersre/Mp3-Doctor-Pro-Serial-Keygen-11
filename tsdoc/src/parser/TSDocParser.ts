// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TextRange } from './TextRange';
import { ParserContext } from './ParserContext';
import { LineExtractor } from './LineExtractor';
import { Tokenizer } from './Tokenizer';
import { NodeParser } from './NodeParser';
import { TSDocConfiguration } from '../configuration/TSDocConfiguration';
import { ParagraphSplitter } from './ParagraphSplitter';

/**
 * The main API for parsing TSDoc comments.
 */
export class TSDocParser {
  /**
   * The configuration that was provided for the TSDocParser.
   */
  public readonly configuration: TSDocConfiguration;

  public constructor(configuration?: TSDocConfiguration) {
    if (configuration) {
      this.configuration = configuration;
    } else {
      this.configuration = new TSDocConfiguration();
    }
  }

  public parseString(text: string): ParserContext {
    return this.parseRange(TextRange.fromString(text));
  }

  public parseRange(range: TextRange): ParserContext {
    const parserContext: ParserContext = new ParserContext(this.configuration, range);

    if (LineExtractor.extract(parserContext)) {
      parserContext.tokens = Tokenizer.readTokens(parserContext.lines);

      const nodeParser: NodeParser = new NodeParser(parserContext);
      nodeParser.parse();

      ParagraphSplitter.splitParagraphs(parserContext.docComment);
    }

    return parserContext;
  }
}
