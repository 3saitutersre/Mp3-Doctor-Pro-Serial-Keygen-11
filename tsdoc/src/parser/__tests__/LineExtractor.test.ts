// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TSDocParser } from '../TSDocParser';
import { TestHelpers } from './TestHelpers';
import type { ParserContext } from '../ParserContext';

function parseAndMatchSnapshot(buffer: string): void {
  const tsdocParser: TSDocParser = new TSDocParser();
  const parserContext: ParserContext = tsdocParser.parseString(buffer);
  expect({
    buffer: TestHelpers.getEscaped(buffer),
    comment: TestHelpers.getEscaped(parserContext.commentRange.toString()),
    lines: parserContext.lines.map((line) => TestHelpers.getEscaped(line.toString())),
    logMessages: parserContext.log.messages.map((message) => message.text)
  }).toMatchSnapshot();
}

test('A. Whitespace variations', () => {
  parseAndMatchSnapshot(`/***/`); // 1
  parseAndMatchSnapshot(` /***/ `); // 2
  parseAndMatchSnapshot(` /** */ `); // 3
  parseAndMatchSnapshot(` /**\n\n*/ `); // 4
  parseAndMatchSnapshot(` /**L1*/ `); // 5
  parseAndMatchSnapshot(` /** L1 */ `); // 6
  parseAndMatchSnapshot(` /**L1\n*/ `); // 7
  parseAndMatchSnapshot(` /**L1*\n*/ `); // 8
  parseAndMatchSnapshot(` /**\nL1*/ `); // 9
  parseAndMatchSnapshot(` /**\n L1 */ `); // 10
  parseAndMatchSnapshot(` /**\nL1\n*/ `); // 11
  parseAndMatchSnapshot(` /**\nL1\n\nL2*/ `); // 12
  parseAndMatchSnapshot(` /**\n*L1\n*/ `); // 13
  parseAndMatchSnapshot(` /**\n * L1\n*/ `); // 14
  parseAndMatchSnapshot(` /**\n * L1\n */ `); // 15
  parseAndMatchSnapshot(` /**L1\n *L2\nL3*/ `); // 16
  parseAndMatchSnapshot(` /** L1\n * L2\n L3*/ `); // 17
  parseAndMatchSnapshot(` /** L1 \n * L2 \n L3 */ `); // 18
  parseAndMatchSnapshot(
    [
      // 19
      '/**  L1  ',
      ' *  L2  ',
      '  L3  ',
      '  L4 */'
    ].join('\r\n')
  );
});

// TODO: Special handling for these somewhat common ornamentations
test('B. Extra stars', () => {
  parseAndMatchSnapshot(` /****/ `);
  parseAndMatchSnapshot(` /**L1**/ `);
  parseAndMatchSnapshot(` /***L1*/ `);
  parseAndMatchSnapshot(`
/*****
 **X**
 *****/ `);
});

test('C. Missing stars', () => {
  parseAndMatchSnapshot(['/**', '```', 'a', ' b', ' c ', '  d', '```', ' */'].join('\n'));

  parseAndMatchSnapshot(['/**', '```', 'ee', ' ff', ' gg ', '  hh', '```', ' */'].join('\n'));
});

test('D. Newline styles', () => {
  parseAndMatchSnapshot(['', '/**', ' * L1', ' */', ''].join('\r\n'));
  parseAndMatchSnapshot(['/**', 'L1', 'L2', '*/'].join('\r\n'));

  // We currently don't support CR or LFCR, so a single "\r" is treated
  // as part of the line.
  parseAndMatchSnapshot(`/** L \r 1 */`);
});

test('E. Parser errors', () => {
  parseAndMatchSnapshot('');
  parseAndMatchSnapshot('/*');
  parseAndMatchSnapshot('//');
  parseAndMatchSnapshot('/** L1\n L2');
  parseAndMatchSnapshot('/** L1 *');
});
