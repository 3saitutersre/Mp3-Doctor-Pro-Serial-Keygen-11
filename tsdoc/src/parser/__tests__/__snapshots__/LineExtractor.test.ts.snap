// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A. Whitespace variations 1`] = `
Object {
  "buffer": "/***/",
  "comment": "/***/",
  "lines": Array [],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 2`] = `
Object {
  "buffer": " /***/ ",
  "comment": "/***/",
  "lines": Array [],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 3`] = `
Object {
  "buffer": " /** */ ",
  "comment": "/** */",
  "lines": Array [],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 4`] = `
Object {
  "buffer": " /**[n][n]*/ ",
  "comment": "/**[n][n]*/",
  "lines": Array [
    "",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 5`] = `
Object {
  "buffer": " /**L1*/ ",
  "comment": "/**L1*/",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 6`] = `
Object {
  "buffer": " /** L1 */ ",
  "comment": "/** L1 */",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 7`] = `
Object {
  "buffer": " /**L1[n]*/ ",
  "comment": "/**L1[n]*/",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 8`] = `
Object {
  "buffer": " /**L1*[n]*/ ",
  "comment": "/**L1*[n]*/",
  "lines": Array [
    "L1*",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 9`] = `
Object {
  "buffer": " /**[n]L1*/ ",
  "comment": "/**[n]L1*/",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 10`] = `
Object {
  "buffer": " /**[n] L1 */ ",
  "comment": "/**[n] L1 */",
  "lines": Array [
    " L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 11`] = `
Object {
  "buffer": " /**[n]L1[n]*/ ",
  "comment": "/**[n]L1[n]*/",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 12`] = `
Object {
  "buffer": " /**[n]L1[n][n]L2*/ ",
  "comment": "/**[n]L1[n][n]L2*/",
  "lines": Array [
    "L1",
    "",
    "L2",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 13`] = `
Object {
  "buffer": " /**[n]*L1[n]*/ ",
  "comment": "/**[n]*L1[n]*/",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 14`] = `
Object {
  "buffer": " /**[n] * L1[n]*/ ",
  "comment": "/**[n] * L1[n]*/",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 15`] = `
Object {
  "buffer": " /**[n] * L1[n] */ ",
  "comment": "/**[n] * L1[n] */",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 16`] = `
Object {
  "buffer": " /**L1[n] *L2[n]L3*/ ",
  "comment": "/**L1[n] *L2[n]L3*/",
  "lines": Array [
    "L1",
    "L2",
    "L3",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 17`] = `
Object {
  "buffer": " /** L1[n] * L2[n] L3*/ ",
  "comment": "/** L1[n] * L2[n] L3*/",
  "lines": Array [
    "L1",
    "L2",
    " L3",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 18`] = `
Object {
  "buffer": " /** L1 [n] * L2 [n] L3 */ ",
  "comment": "/** L1 [n] * L2 [n] L3 */",
  "lines": Array [
    "L1",
    "L2",
    " L3",
  ],
  "logMessages": Array [],
}
`;

exports[`A. Whitespace variations 19`] = `
Object {
  "buffer": "/**  L1  [r][n] *  L2  [r][n]  L3  [r][n]  L4 */",
  "comment": "/**  L1  [r][n] *  L2  [r][n]  L3  [r][n]  L4 */",
  "lines": Array [
    " L1",
    " L2",
    "  L3",
    "  L4",
  ],
  "logMessages": Array [],
}
`;

exports[`B. Extra stars 1`] = `
Object {
  "buffer": " /****/ ",
  "comment": "/****/",
  "lines": Array [
    "*",
  ],
  "logMessages": Array [],
}
`;

exports[`B. Extra stars 2`] = `
Object {
  "buffer": " /**L1**/ ",
  "comment": "/**L1**/",
  "lines": Array [
    "L1*",
  ],
  "logMessages": Array [],
}
`;

exports[`B. Extra stars 3`] = `
Object {
  "buffer": " /***L1*/ ",
  "comment": "/***L1*/",
  "lines": Array [
    "*L1",
  ],
  "logMessages": Array [],
}
`;

exports[`B. Extra stars 4`] = `
Object {
  "buffer": "[n]/*****[n] **X**[n] *****/ ",
  "comment": "/*****[n] **X**[n] *****/",
  "lines": Array [
    "***",
    "*X**",
    "***",
  ],
  "logMessages": Array [],
}
`;

exports[`C. Missing stars 1`] = `
Object {
  "buffer": "/**[n][c][c][c][n]a[n] b[n] c [n]  d[n][c][c][c][n] */",
  "comment": "/**[n][c][c][c][n]a[n] b[n] c [n]  d[n][c][c][c][n] */",
  "lines": Array [
    "[c][c][c]",
    "",
    " ",
    " ",
    "  ",
    "[c][c][c]",
  ],
  "logMessages": Array [],
}
`;

exports[`C. Missing stars 2`] = `
Object {
  "buffer": "/**[n][c][c][c][n]ee[n] ff[n] gg [n]  hh[n][c][c][c][n] */",
  "comment": "/**[n][c][c][c][n]ee[n] ff[n] gg [n]  hh[n][c][c][c][n] */",
  "lines": Array [
    "[c][c][c]",
    "ee",
    " ff",
    " gg",
    "  hh",
    "[c][c][c]",
  ],
  "logMessages": Array [],
}
`;

exports[`D. Newline styles 1`] = `
Object {
  "buffer": "[r][n]/**[r][n] * L1[r][n] */[r][n]",
  "comment": "/**[r][n] * L1[r][n] */",
  "lines": Array [
    "L1",
  ],
  "logMessages": Array [],
}
`;

exports[`D. Newline styles 2`] = `
Object {
  "buffer": "/**[r][n]L1[r][n]L2[r][n]*/",
  "comment": "/**[r][n]L1[r][n]L2[r][n]*/",
  "lines": Array [
    "L1",
    "L2",
  ],
  "logMessages": Array [],
}
`;

exports[`D. Newline styles 3`] = `
Object {
  "buffer": "/** L [r] 1 */",
  "comment": "/** L [r] 1 */",
  "lines": Array [
    "L [r] 1",
  ],
  "logMessages": Array [],
}
`;

exports[`E. Parser errors 1`] = `
Object {
  "buffer": "",
  "comment": "",
  "lines": Array [],
  "logMessages": Array [
    "Expecting a \\"/**\\" comment",
  ],
}
`;

exports[`E. Parser errors 2`] = `
Object {
  "buffer": "/*",
  "comment": "",
  "lines": Array [],
  "logMessages": Array [
    "(1,1): Expecting a \\"/**\\" comment",
  ],
}
`;

exports[`E. Parser errors 3`] = `
Object {
  "buffer": "//",
  "comment": "",
  "lines": Array [],
  "logMessages": Array [
    "(1,1): Expecting a leading \\"/**\\"",
  ],
}
`;

exports[`E. Parser errors 4`] = `
Object {
  "buffer": "/** L1[n] L2",
  "comment": "",
  "lines": Array [],
  "logMessages": Array [
    "(1,1): Unexpected end of input",
  ],
}
`;

exports[`E. Parser errors 5`] = `
Object {
  "buffer": "/** L1 *",
  "comment": "",
  "lines": Array [],
  "logMessages": Array [
    "(1,1): Unexpected end of input",
  ],
}
`;
